### ISSUE TITLE
3.14.x: Prerelease publish failed to bump `version` in most `package.json` files.

### ISSUE THREAD

**Issue by abernix** [lerna/lerna#2107]
This behavior is a scripted behavior which we've used for some time now, but seems to have broken with 3.14.1 (presumably `3.14.0` too, but we'd been using `3.13.4` before).  Essentially, Lerna didn't bump the `version`s for the packages it said it was going to bump, committed the one that it did update (the very first one), and then went on to try and publish — ultimately failing because it attempts to publish over the previously published version.

## Expected Behavior

When running the following, I expect it to suggest bumping all the appropriate packages by a prerelease faction, prompt with proposed versions, and when confirmed, publish them to their appropriate `prerelease` versions (e.g. `.alpha.n`, in this case):

```
lerna publish --exact \
	--force-publish=apollo-server-core \
	--include-merged-tags \
	--preid alpha \
	--dist-tag next \
	prerelease 
```

As I expected, this yielded a prompt of suggested version bumps, as it typically does:

```
Changes:
 - apollo-datasource-rest: 0.4.0 => 0.4.1-alpha.0
 - apollo-engine-reporting: 1.2.0-alpha.3 => 1.2.0-alpha.4
 - apollo-server-azure-functions: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-cloud-functions: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-cloudflare: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-core: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-express: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-fastify: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-hapi: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-integration-testsuite: 2.6.0-alpha.3 => 2.6.0-alpha.4 (private)
 - apollo-server-koa: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-lambda: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-micro: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server-plugin-base: 0.5.0-alpha.3 => 0.5.0-alpha.4
 - apollo-server-testing: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - apollo-server: 2.6.0-alpha.3 => 2.6.0-alpha.4
 - graphql-extensions: 0.7.0-alpha.3 => 0.7.0-alpha.4
```

Two of these packages have relevancy to this bug, but I've included the full output for accuracy.  Of note are:

```
apollo-datasource-rest: 0.4.0 => 0.4.1-alpha.0
```

and

```
apollo-server-fastify: 2.6.0-alpha.3 => 2.6.0-alpha.4
```

Those will be relevant in the next section.

## Current Behavior

The above typically works, publishing the proposed packages accordingly (barring registry failures), but it appears that some characteristics that I believe are new to `3.14.x` prevented it from updating all of the packages and it instead only committed (to Git) the `package.json` bump to the `version` property for the very first package (in this case, `apollo-datasource-rest`), and then proceeded to attempt to publish the other packages _without_ bumping the `version` in their `package.json`'s, resulting in:

```
? Are you sure you want to publish these packages? Yes
lerna info execute Skipping GitHub releases
lerna info git Pushing tags...
lerna info publish Publishing packages to npm...
lerna info Verifying npm credentials
lerna http fetch GET 200 https://registry.npmjs.org/-/npm/v1/user 598ms
lerna http fetch GET 200 https://registry.npmjs.org/-/org/apollo-bot/package?format=cli 350ms
lerna http fetch PUT 403 https://registry.npmjs.org/apollo-server-fastify 569ms
lerna ERR! E403 You cannot publish over the previously published versions: 2.6.0-alpha.3
```

What's happened here is that it's tried publishing `apollo-server-fastify@2.6.0-alpha.3` (`.3`), even though it said it was going to bump to and publish `apollo-server-fastify@2.6.0-alpha.4` (`.4`).

Digging into the aftermath, I had the following commit on the tip of my branch (commit body expand-o'd in this `<detail>` for brevity):

<details><summary>The `Publish` commit</summary><p>

```diff
commit d25304e123f9176e1980e59e03ac57b03cf76f98
Author: Jesse Rosenberger <email@domain.cc>
Date:   Thu May 23 19:30:10 2019 +0300

    Publish
    
     - apollo-datasource-rest@0.4.1-alpha.0
     - apollo-engine-reporting@1.2.0-alpha.4
     - apollo-server-azure-functions@2.6.0-alpha.4
     - apollo-server-cloud-functions@2.6.0-alpha.4
     - apollo-server-cloudflare@2.6.0-alpha.4
     - apollo-server-core@2.6.0-alpha.4
     - apollo-server-express@2.6.0-alpha.4
     - apollo-server-fastify@2.6.0-alpha.4
     - apollo-server-hapi@2.6.0-alpha.4
     - apollo-server-integration-testsuite@2.6.0-alpha.4
     - apollo-server-koa@2.6.0-alpha.4
     - apollo-server-lambda@2.6.0-alpha.4
     - apollo-server-micro@2.6.0-alpha.4
     - apollo-server-plugin-base@0.5.0-alpha.4
     - apollo-server-testing@2.6.0-alpha.4
     - apollo-server@2.6.0-alpha.4
     - graphql-extensions@0.7.0-alpha.4

diff --git a/packages/apollo-datasource-rest/package.json b/packages/apollo-datasource-rest/package.json
index 562d4c6b..6a7cfeb0 100644
--- a/packages/apollo-datasource-rest/package.json
+++ b/packages/apollo-datasource-rest/package.json
@@ -1,6 +1,6 @@
 {
   "name": "apollo-datasource-rest",
-  "version": "0.4.0",
+  "version": "0.4.1-alpha.0",
   "author": "not-leaving-my-email-here@domain.com",
   "license": "MIT",
   "repository": {
```
</p></details>

Most of the version bumps are missing here, and it would have been expected to bump the other packages it thought it wanted to bump as well.

It had, however, gone ahead and `git tag`ged all the packages though (snipped some tags for brevity, again):

```
(HEAD -> release-2.6.0, tag: graphql-extensions@0.7.0-alpha.4, tag: apollo-server@2.6.0-alpha.4, ... tag: apollo-server-fastify@2.6.0-alpha.4, ...  tag: apollo-datasource-rest@0.4.1-alpha.0, origin/release-2.6.0)
```

And ultimately, it had published nothing.  (I consider this a good thing, given its likely doomed fate!)

## Steps to Reproduce (for bugs)

1.   
   ``` 
   lerna publish --exact \
	 --force-publish=apollo-server-core \
	 --include-merged-tags \
	--preid alpha \
	--dist-tag next \
	prerelease 
   ```

<details><summary>lerna.json</summary><p>

```json
{
  "packages": ["packages/*"],
  "version": "independent"
}
```
</p></details>

## Context

I realize that undoing a failure here is difficult and there might be a better way to undo the failure, but in this case, I wasn't sure as to the best approach.  Typically, if it does the bumps and fails in an `npm` step, I'm still able to use `from-git` to use the committed details.

In this case, however, the commit was suggesting that a bunch of packages had been updated in its message but the versions were only bumped for one.  I therefore deleted the new `git tag`s both locally and on the remote server, my working tree to the previous (pre-`Publish`) commit via `git reset --hard HEAD~` and tried the publish again.

Ultimately, I reverted to `lerna@3.13.4`, tried again, and had success.

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.14.1 |
| `npm --version`  | 6.9.0 |
| `node --version` | 10.15.1 |

| OS | Version |
| --- | --- |
| NAME | VERSION |
| macOS Mojave | 10.14.4 |

Happy to try and help however!

**Comment by evocateur** [comment]
Do you have any lifecycle scripts in the root or leaves that `git checkout ...` anywhere? If they're run in `version` or `postversion`, I believe those will pollute (that is, break) the collection of proper changes.

In any case, I would very much appreciate a minimal reproduction to make progress on a diagnosis.

**Comment by nicolo-ribaudo** [comment]
We have this same issue in Babel, but with a "normal" version (non a prerelease).

Version 7.5.0 had this problem; I had to manually update all the `package.json`s
Version 7.5.1 to 7.5.4 worked fine.
Version 7.5.5 (which I want to release) has this same problem.

One thing that I noticed is that the `package.json` files which have been updated in the upcoming 7.5.5 and in 7.5.0 are the same.

I'm really sorry, but I'm going to give exactly the opposite of a _minimal_ example. :sweat_smile:  I'll also try to debug it and report anything I find.

Steps to reproduce:
1. Fork https://github.com/babel/babel/
2. `git reset --hard d8da63c929f2d28c401571e2a43166678c555bc4` (and, if present, `git tag --delete v7.5.5`)
3. `git push -f`, to also reset `origin` (and, if present, `git push --delete origin v7.5.5`)
4. `./node_modules/.bin/lerna version --force-publish="@babel/runtime,@babel/runtime-corejs2,@babel/runtime-corejs3,@babel/standalone,@babel/preset-env-standalone"`

We don't have any `version` or `postversion` script.

**Comment by danez** [comment]
I think the issue lies in this function https://github.com/lerna/lerna/blob/master/utils/query-graph/query-graph.js#L62 which was introduced in 3.14.0

If I understand it correctly it gets all the leaf packages first (`this._getNextLeaf()`). If there are non left it tries to get the cycle packages BUT only if all that is left are cycle packages (`this._onlyCyclesLeft()`).

So the case we (babel) are running into is that after all the leaf packages are done, we have at least  one package that is not a leaf, but it is also not part of any cycle (e.g. `@babel/helper-replace-supers`). In that case the function returns an empty array and I think at this point lerna stops updating any further packages and moves one to committing, thus leaving out a big part of our packages.
But I just did a short debug session, I might have missed something though.

**Comment by nicolo-ribaudo** [comment]
I was going to write the same conclusion

**Comment by bweggersen** [comment]
Great debugging! I was able to repro the bug through a new test case. I'll look into it and create a PR with a fix. Thanks everyone!

**Comment by nicolo-ribaudo** [comment]
@bweggersen I was also working on a PR :stuck_out_tongue:
Btw, I think that it can still be made a little faster, because some "leaf cycles" can be executed in parallel with leaf nodes.

**Comment by bweggersen** [comment]
Oh, I should have checked first! 🙃

I'm happy to let you drive the PR since you have pretty ASCII art :) Looking into the problem, this is what I came up with: https://github.com/lerna/lerna/compare/master...bweggersen:bweggersen/cycle-with-common-parent

For transparency and clarity, this is my understanding of the order of execution (based on existing lerna functionality):
If there is an available leaf node, pick it. If not, and if we have cyclical nodes, then pick one of those ([ref](https://github.com/lerna/lerna/blob/master/commands/run/__tests__/run-command.test.js#L231)). Only start working on a cyclical node if we only have cyclical nodes left. As far as I understand, we have this constraint to make sure we don't start working on cyclical nodes while there are unbuilt dependencies of those cyclical nodes. It is possible that cyclical nodes are grandparents to nodes currently being worked on. So those nodes, and their parents, should be completed first. However, in situations like you illustrate in Edit 3, after having worked on B we have no available leaf node, but C and D are not alone either. We have an ordinary node in A. On [my branch](https://github.com/lerna/lerna/compare/master...bweggersen:bweggersen/cycle-with-common-parent#diff-c986cbd00c661304b587a380d26f2b03), I added a way to track tasks in progress. If we don't have any available leaf nodes, but we're not working on any tasks either, we can start picking up cyclical tasks, even if we have other ordinary nodes further up in the graph.

@evocateur Any feedback on what you think is the right thing to do here?

**Comment by nicolo-ribaudo** [comment]
I am following an approach similar to this:

1. When looking for cycles, collaps them to a single node. In the Edit 3 of https://github.com/lerna/lerna/issues/2107#issuecomment-511987691, C-D would be collapsed to a single CD node without dependencies and whose only dependent is A.
2. CD and B are both leaf nodes without dependencies; they can be executed.
3. After both CD and B are done; A doesn't have any dependency left and can be executed.

I'm currently trying to make it work with nested cycles and cycles with intersection.

**Comment by evocateur** [comment]
Honestly, all the graph traversal stuff is _extremely_ out of my wheelhouse. I'm very appreciative of the expert help here.

@bweggersen I'm reviewing @nicolo-ribaudo's PR right now, I can take a look at your branch after my head stops spinning...

**Comment by nicolo-ribaudo** [comment]
Our approaches are completely different: I think that with my approach lerna respects the topological order better, but @bweggersen is way simpler.

**Comment by evocateur** [comment]
@nicolo-ribaudo I really like your PR's improvements to the cyclic logging itself. Way more readable/actionable!

**Comment by evocateur** [comment]
Fix is in [v3.16.0](https://github.com/lerna/lerna/releases/tag/v3.16.0), just released

**Comment by abernix** [comment]
Immense gratitude to those (especially @nicolo-ribaudo with the PR) who were able to jump in and get this resolved!

❤️
